src/Microsoft.OData.Core/ODataCollectionWriterCore.cs

```diff
@@ -508,20 +511,52 @@ private void StartPayloadInStartState()
             Scope current = this.scopes.Peek();
             if (current.State == CollectionWriterState.Start)
             {
-                this.InterceptException(this.StartPayload);
+                this.InterceptException((thisParam) => thisParam.StartPayload());
             }
         }
 
         /// <summary>
         /// Catch any exception thrown by the action passed in; in the exception case move the writer into
-        /// state ExceptionThrown and then rethrow the exception.
+        /// state Error and then re-throw the exception.
         /// </summary>
         /// <param name="action">The action to execute.</param>
-        private void InterceptException(Action action)
+        /// <remarks>
+        /// Make sure to only use anonymous functions that don't capture state from the enclosing context, 
+        /// so the compiler optimizes the code to avoid creating allocations on every call to this method.
+        /// </remarks>
+        private void InterceptException(Action<ODataCollectionWriterCore> action)
+        {
+            try
+            {
+                action(this);

```

Comment:
Maybe it would be a good idea to mention that this is done as an optimization to avoid delegate allocations?